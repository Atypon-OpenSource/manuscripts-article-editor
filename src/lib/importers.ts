import JSZip from 'jszip'
import { extname } from 'path'
import { FIGURE } from '../transformer/object-types'
import {
  Attachments,
  ContainedModel,
  Figure,
  Model,
  ModelAttachment,
} from '../types/models'
import { generateAttachmentFilename } from './exporter'
import { convert } from './pressroom'

type Importer = (file: File) => Promise<Model[]>

interface Importers {
  [key: string]: Importer
}

export interface JsonModel
  extends ContainedModel,
    Attachments,
    ModelAttachment {
  _id: string
  bundled: boolean
  collection: string
  contentType: string
}

export interface ProjectDump {
  version: string
  data: JsonModel[]
}

const modelHasObjectType = <T extends Model>(objectType: string) => (
  model: Model
): model is T => {
  return model.objectType === objectType
}

export const readManuscriptFromBundle = async (
  zip: JSZip
): Promise<ProjectDump> => {
  const json = await zip.file('index.manuscript-json').async('text')

  return JSON.parse(json)
}

const importProjectBundle = async (result: Blob) => {
  const zip = await new JSZip().loadAsync(result)

  const doc = await readManuscriptFromBundle(zip)

  if (doc.version !== '2.0') {
    throw new Error(`Unsupported version: ${doc.version}`)
  }

  // TODO: validate?
  // TODO: ensure default data is added

  const items = doc.data
    .filter(item => !item.bundled)
    .filter(item => item._id)
    .map(item => {
      delete item._rev
      delete item.collection

      return item
    })

  const folder = zip.folder('Data')

  await Promise.all(
    items
      .filter(modelHasObjectType<Figure>(FIGURE))
      .map(async (item: Figure) => {
        const filename = generateAttachmentFilename(item._id)

        try {
          item.attachment = {
            id: filename, // TODO: original name?
            type: item.contentType || 'image/png',
            data: await folder.file(filename).async('blob'),
          }
        } catch (error) {
          // tslint:disable-next-line:no-console
          console.error(`Could not retrieve attachment from zip: ${error}`)
          // continue without attachment
        }
      })
  )

  return items
}

const importConvertedFile: Importer = async (file: File) => {
  const form = new FormData()
  form.append('file', file)

  const result = await convert(form, '.manuproj')

  // download(result, 'manuscript.zip')

  return importProjectBundle(result)
}

const importers: Importers = {
  '.docx': importConvertedFile,
  '.html': importConvertedFile,
  '.md': importConvertedFile,
  '.manuproj': importProjectBundle,
}

export const openFilePicker = (): Promise<File> =>
  new Promise((resolve, reject) => {
    const input = document.createElement('input')
    input.type = 'file'
    input.accept = Object.keys(importers).join(',')
    input.addEventListener('change', () => {
      if (input.files && input.files.length) {
        resolve(input.files[0])
      } else {
        reject(new Error('No file was received'))
      }
    })
    input.click()
  })

export const importFile = async (file: File) => {
  const extension = extname(file.name)

  return importers[extension](file)
}
