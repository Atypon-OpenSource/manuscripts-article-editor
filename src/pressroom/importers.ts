/*!
 * The contents of this file are subject to the Common Public Attribution License Version 1.0 (the “License”); you may not use this file except in compliance with the License. You may obtain a copy of the License at https://mpapp-public.gitlab.io/manuscripts-frontend/LICENSE. The License is based on the Mozilla Public License Version 1.1 but Sections 14 and 15 have been added to cover use of software over a computer network and provide for limited attribution for the Original Developer. In addition, Exhibit A has been modified to be consistent with Exhibit B.
 *
 * Software distributed under the License is distributed on an “AS IS” basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the specific language governing rights and limitations under the License.
 *
 * The Original Code is manuscripts-frontend.
 *
 * The Original Developer is the Initial Developer. The Initial Developer of the Original Code is Atypon Systems LLC.
 *
 * All portions of the code written by Atypon Systems LLC are Copyright (c) 2019 Atypon Systems LLC. All Rights Reserved.
 */

import {
  ModelAttachment,
  parseJATSArticle,
  parseSTSStandard,
} from '@manuscripts/manuscript-transform'
import {
  Bundle,
  ContributorRole,
  Model,
  ObjectTypes,
} from '@manuscripts/manuscripts-json-schema'
import JSZip from 'jszip'
import { flatMap } from 'lodash-es'
import { basename, extname } from 'path'
import { FileExtensionError } from '../lib/errors'
import {
  createNewContributorRoles,
  createNewStyles,
  fetchSharedData,
} from '../lib/templates'
import { cleanItem } from './clean-item'
import { removeUnsupportedData } from './exporter'
import { convert } from './pressroom'

export interface JsonModel extends Model, ModelAttachment {
  _id: string
  bundled?: boolean
  collection?: string
  contentType?: string
}

export interface ProjectDump {
  version: string
  data: JsonModel[]
}

export const readProjectDumpFromArchive = async (
  zip: JSZip
): Promise<ProjectDump> => {
  const json = await zip.file('index.manuscript-json').async('text')

  return JSON.parse(json)
}

const attachmentKeys: { [key in ObjectTypes]?: string } = {
  [ObjectTypes.Figure]: 'image',
  [ObjectTypes.Bundle]: 'csl',
}

const defaultAttachmentContentTypes: { [key in ObjectTypes]?: string } = {
  [ObjectTypes.Figure]: 'image/png',
  [ObjectTypes.Bundle]: 'application/vnd.citationstyles.style+xml',
}

const importProjectArchive = async (result: Blob) => {
  const zip = await new JSZip().loadAsync(result)

  const projectDump = await readProjectDumpFromArchive(zip)

  if (projectDump.version !== '2.0') {
    throw new Error(`Unsupported version: ${projectDump.version}`)
  }

  // TODO: validate?
  // TODO: ensure default data is added
  // TODO: add default bundle (which has no parent bundle)
  // TODO: ensure that pageLayout and bundle are set

  const items = projectDump.data
    .filter(item => !item.bundled)
    .filter(item => item.objectType !== 'MPContentSummary')
    .filter(item => item._id)
    .map(item => cleanItem(item))

  // load attachments from the Data folder

  for (const file of Object.values(zip.files)) {
    if (!file.dir && file.name.startsWith('Data/')) {
      const id = basename(file.name).replace('_', ':')

      const model = items.find(model => model._id === id)

      if (model) {
        const objectType = model.objectType as ObjectTypes

        if (objectType in attachmentKeys) {
          model.attachment = {
            id: attachmentKeys[objectType]!,
            type:
              model.contentType || defaultAttachmentContentTypes[objectType]!,
            data: await file.async('blob'),
          }
        }
      }
    }
  }

  return items
}

interface BundledData {
  bundles: Map<string, Bundle>
  contributorRoles: Map<string, ContributorRole>
  styles: Map<string, Model>
}

export const importBundledData = async (): Promise<BundledData> => {
  const bundles = await fetchSharedData<Bundle>('bundles')
  const contributorRoles = await fetchSharedData<ContributorRole>(
    'contributor-roles'
  )
  const styles = await fetchSharedData<Model>('styles')

  return {
    bundles,
    contributorRoles: createNewContributorRoles(contributorRoles),
    styles: createNewStyles(styles),
  }
}

const convertFile = async (file: File): Promise<Blob> => {
  const form = new FormData()
  form.append('file', file)

  return convert(form, '.manuproj')
}

const parseXMLFile = async (file: File): Promise<Document> => {
  const url = window.URL.createObjectURL(file)
  const text = await fetch(url).then(response => response.text())
  return new DOMParser().parseFromString(text, 'application/xml')
}

const convertXMLDocument = (doc: Document): Model[] => {
  // TODO: check doc.doctype.publicId?
  switch (doc.documentElement.nodeName) {
    case 'article':
      return parseJATSArticle(doc)

    case 'standard':
      return parseSTSStandard(doc)

    default:
      throw new Error('Unsupported XML format')
  }
}

interface FileType {
  extension: string
  mimetypes: string[]
  description: string
}

const fileTypes: FileType[] = [
  {
    extension: '.docx',
    mimetypes: [
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    ],
    description: 'DOCX (Microsoft Word)',
  },
  {
    extension: '.doc',
    mimetypes: ['application/msword'],
    description: 'DOC (Microsoft Word)',
  },
  {
    extension: '.html',
    mimetypes: ['text/html'],
    description: 'HTML',
  },
  {
    extension: '.md',
    mimetypes: ['text/markdown', 'text/plain'],
    description: 'Markdown',
  },
  {
    extension: '.manuproj',
    mimetypes: ['application/zip'],
    description: 'Manuscripts Project Bundle',
  },
  {
    extension: '.latex',
    mimetypes: [
      'application/x-latex',
      'application/latex',
      'text/x-latex',
      'text/latex',
      'text/plain',
    ],
    description: 'LaTeX',
  },
  {
    extension: '.tex',
    mimetypes: [
      'application/x-tex',
      'application/tex',
      'text/x-tex',
      'text/tex',
      'text/plain',
    ],
    description: 'TeX',
  },
  {
    extension: '.xml',
    mimetypes: ['application/xml', 'text/xml'],
    description: 'JATS/STS XML',
  },
  {
    extension: '.zip',
    mimetypes: ['application/zip'],
    description: 'ZIP (containing Markdown or LaTeX)', // TODO: could also be XML or HTML
  },
]

export const acceptedFileExtensions = () => {
  return fileTypes.map(item => item.extension)
}

export const acceptedFileDescription = () => {
  return fileTypes.map(item => item.description)
}

export const acceptedMimeTypes = () => {
  return flatMap(fileTypes, item => item.mimetypes)
}

export const openFilePicker = (
  acceptedExtensions: string[] = acceptedFileExtensions()
): Promise<File> =>
  new Promise((resolve, reject) => {
    const input = document.createElement('input')
    input.type = 'file'
    input.accept = acceptedExtensions.join(',')

    const handleFocus = () => {
      window.removeEventListener('focus', handleFocus)

      // This event is fired before the input's change event,
      // and before the input's FileList has been populated,
      // so a delay is needed.
      window.setTimeout(() => {
        if (!input.files || !input.files.length) {
          resolve()
        }
      }, 1000)
    }

    // window "focus" event, fired even if the file picker is cancelled.
    window.addEventListener('focus', handleFocus)

    input.addEventListener('change', () => {
      if (input.files && input.files.length) {
        const [file] = input.files

        const extension = extname(file.name)

        if (acceptedExtensions.includes(extension)) {
          resolve(file)
        } else {
          const error = new FileExtensionError(extension, acceptedExtensions)
          reject(error)
        }
      } else {
        resolve()
      }
    })

    input.click()
  })

export const importFile = async (file: File) => {
  const { name, lastModified } = file

  if (name.endsWith('.manuproj')) {
    const zip = await new JSZip().loadAsync(file)
    await removeUnsupportedData(zip)
    const blob = await zip.generateAsync({ type: 'blob' })
    file = new File([blob], name, { lastModified })
  }

  // TODO: look inside .zip files
  if (name.endsWith('.xml')) {
    return parseXMLFile(file).then(convertXMLDocument)
  }

  const result = await convertFile(file)

  return importProjectArchive(result)
}
