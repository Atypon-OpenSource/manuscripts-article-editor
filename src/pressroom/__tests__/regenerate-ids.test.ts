/*!
 * The contents of this file are subject to the Common Public Attribution License Version 1.0 (the “License”); you may not use this file except in compliance with the License. You may obtain a copy of the License at https://mpapp-public.gitlab.io/manuscripts-frontend/LICENSE. The License is based on the Mozilla Public License Version 1.1 but Sections 14 and 15 have been added to cover use of software over a computer network and provide for limited attribution for the Original Developer. In addition, Exhibit A has been modified to be consistent with Exhibit B.
 *
 * Software distributed under the License is distributed on an “AS IS” basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for the specific language governing rights and limitations under the License.
 *
 * The Original Code is manuscripts-frontend.
 *
 * The Original Developer is the Initial Developer. The Initial Developer of the Original Code is Atypon Systems LLC.
 *
 * All portions of the code written by Atypon Systems LLC are Copyright (c) 2020 Atypon Systems LLC. All Rights Reserved.
 */

import fs from 'fs'
import JSZip from 'jszip'

import {
  updateAttachments,
  updateIdentifiers,
} from '../../lib/update-identifiers'
import { readProjectDumpFromArchive } from '../importers'

const buffer = fs.readFileSync(__dirname + '/__fixtures__/example.manuproj')
const blob = new Blob([buffer])

describe('regenerate ids', () => {
  test('regenerates all ids', async () => {
    const zip = await new JSZip().loadAsync(blob)
    const { data } = await readProjectDumpFromArchive(zip)

    const idMap = await updateIdentifiers(data)

    expect(idMap.size).toBe(131)

    const keys = [...idMap.keys()]
    expect(keys).toMatchSnapshot()

    const values = [...idMap.values()]
    expect(values).not.toEqual(keys)
  })

  test('updates attachment file names', async () => {
    const zip = await new JSZip().loadAsync(blob)
    const { data } = await readProjectDumpFromArchive(zip)

    const idMap = await updateIdentifiers(data)
    await updateAttachments(zip, idMap)

    const result = Object.keys(zip.files)

    expect(result).toHaveLength(5)

    expect(result).toContain('index.manuscript-json')
    expect(result).toContain('Data/')
    expect(result).toContain('containers.json')

    expect(result).not.toContain(
      'Data/MPBundle_D01D05F3-9C1F-4424-85C4-817969C0B2BC'
    )
    expect(result).not.toContain(
      'Data/MPFigure_3DD0D874-52DF-4D0B-A39D-A12C19337438'
    )
  })
})
