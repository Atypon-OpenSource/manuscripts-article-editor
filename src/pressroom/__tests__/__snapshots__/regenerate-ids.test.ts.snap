// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`regenerate ids regenerates all ids 1`] = `
Array [
  "MPAuxiliaryObjectReferenceStyle:FF9DEBFE-56AF-41B2-9D0A-0857EA14C8A1",
  "MPStyleableStringComponent:B8CF9955-0956-4114-AAC7-F9C6CC99688F",
  "MPBorderStyle:519CC6EE-6D5D-4EA7-8E80-A6C70F1452AC",
  "MPBorderStyle:7204B4A9-DCF0-4F20-AE93-BE5E227C1B2E",
  "MPBorderStyle:D7318C5F-589C-464A-B70D-5D4E99866D9E",
  "MPBorderStyle:F5D92FF0-368E-4902-960E-51DA46A86C4A",
  "MPBundle:D01D05F3-9C1F-4424-85C4-817969C0B2BC",
  "MPCaptionStyle:0D6BD4E7-3F35-4612-82DE-1F4922A14F60",
  "MPStyleableStringComponent:E43EAFE4-7C81-468E-BF88-8CEF41435062",
  "MPStyleableStringComponent:A6F4C1B4-597F-4206-A21A-4CA60F6C7A2E",
  "MPStyleableStringComponent:E5CB691A-EE39-41BD-8D78-22E02BD7782A",
  "MPStyleableStringComponent:9E4DDC68-4667-49C8-9F3E-600E0258A53A",
  "MPColor:137A174D-6E3E-4F94-89D3-FF217E129F72",
  "MPColor:13E6E476-4F54-432F-BCB9-C02D52887CE4",
  "MPColor:2F61ED90-7E7B-4996-94BB-BFE98A415491",
  "MPColor:369728A0-94E4-47A9-B2AA-E0159E09163A",
  "MPColor:3B4F3FE7-9059-4E52-BDD0-5EA654A1DD07",
  "MPColor:4BA62F0D-ADFA-48EE-96E7-F9C9168EB369",
  "MPColor:4BD7CF21-D98A-4FEC-861D-E99FFE14CBF5",
  "MPColor:5AA5D6C8-C5C7-4ACC-9DDC-BF88E45DD73C",
  "MPColor:5CAF29B2-1574-46A3-984D-E709C07EEEAB",
  "MPColor:61725D61-C989-448B-BD9C-B995443BC5F3",
  "MPColor:742EA4C8-FEB7-4876-A69F-283407D0E086",
  "MPColor:90DB3DF9-4992-4154-BD3E-3A31BC67891D",
  "MPColor:D816843A-8802-4957-B689-004FCB9C5A11",
  "MPColor:DDAD4889-4778-47DB-AC72-75BD675944E0",
  "MPColor:E8F2C41D-0043-4EE5-9BE9-23ECB3B8E255",
  "MPColor:FB247F45-0513-40B2-B3D1-AB3C27ACF92F",
  "MPColorScheme:770903C8-C46C-491D-BAAC-D12F0F65CCF0",
  "MPContributor:22394227-2B83-41A2-A33C-AA74B72328B7",
  "MPBibliographicName:3BFBC390-744B-4A83-B6BD-3A46A3F4D93E",
  "MPContributorRole:0DC27D5C-9D22-4D60-9A20-D1DC5DBC576D",
  "MPContributorRole:334AD77E-87B1-47BF-95EB-347AAD062200",
  "MPContributorRole:61509A44-61AD-499C-8F8C-795167EEFF1C",
  "MPContributorRole:6B0E6FF1-2E81-4047-B49E-568A0E3EFAFF",
  "MPContributorRole:6D93E55D-181C-4E70-8350-5952B2D2ABC9",
  "MPContributorRole:75A5BF6A-8DC1-4989-98E8-8962E7B824F6",
  "MPContributorRole:9EA8B8C8-6C4E-4ED9-9128-6217DCAB01C7",
  "MPContributorRole:A81B5783-4F56-4B38-9340-FBD9F3872701",
  "MPContributorRole:ADD48290-783D-421D-BEEE-FDC036817F05",
  "MPContributorRole:B411429A-28F5-4EFB-ACFF-6A8E0333B54A",
  "MPContributorRole:D0E02ABE-EEFB-4E97-93BC-5A2AE3E47456",
  "MPContributorRole:EE441AE4-38FA-4A77-B323-4107991E8F58",
  "MPContributorRole:F2936D97-BFB0-4839-A661-13BE68F3A5ED",
  "MPContributorRole:FE06C422-AA94-4120-9F76-83F6134B47B1",
  "MPFigureLayout:165FA0EB-E651-4237-958C-733AACB14A65",
  "MPFigureLayout:1E191114-2B5C-49A7-B218-165C0DBB7EFD",
  "MPFigureLayout:2AFEBA4E-9CAE-4EC9-B4BC-28E6F6405639",
  "MPFigureLayout:57AD1367-2F78-4DA9-921B-BDAD87363BD6",
  "MPFigureLayout:B28E2F50-2692-46BF-A3F0-75317312D424",
  "MPFigureLayout:C9FC4AC2-0E6D-45F1-B488-FCBED7181DC4",
  "MPFigureLayout:D30986C8-75D0-41BF-9712-5EB37D0A5D7E",
  "MPFigureLayout:F47C30C2-AEE8-4B0D-B799-462123083918",
  "MPFigureLayout:F6AA7D5C-FC9D-41BD-824D-5C4A957C3A72",
  "MPFigureStyle:9FC55E4B-AF7C-459F-B53A-AD0DD2EB74BF",
  "MPBorder:0FFF4CDC-9D5C-4B7D-A606-F0374354FB7F",
  "MPManuscript:81AD729F-04FE-478C-8B6C-98D07BB64675",
  "MPMaximumSectionWordCountRequirement:1733EAA4-3805-4A67-B61F-D3FCB1D348A1",
  "MPMaximumSectionWordCountRequirement:C0C36279-71B4-4EF4-9BE8-40D89727BD29",
  "MPPageLayout:D1C6D0AA-7F58-4C1A-BEF7-3C09FFB01D88",
  "MPNumberingStyle:CB6FF724-7F91-454A-9A3C-30293471E039",
  "MPParagraphElement:0E47A2B1-6CC6-4129-8908-28A7130A9879",
  "MPParagraphElement:0F53D996-229E-4B7C-818C-0CC1E3249C18",
  "MPParagraphElement:AE9CB377-F253-476D-8422-7A744E614DD4",
  "MPParagraphElement:E2E17921-C161-4399-8E13-7EC1F2B249F0",
  "MPParagraphStyle:01AE40EF-E90C-4366-A762-2066014A086B",
  "MPTextStyling:BBA06E82-F724-419F-A81E-1886644AA4B9",
  "MPParagraphStyle:0F315B3B-FBD5-4893-9950-A0DC9332B648",
  "MPTextStyling:A6E169AC-4610-44C6-A979-7FD95AE787CC",
  "MPParagraphStyle:10C0C6AE-08E5-48AD-B8CC-1A58CD3001BE",
  "MPTextStyling:2E2D93EB-2474-4494-AD92-B5EAD46A38BE",
  "MPParagraphStyle:292B5512-59C2-480E-83A8-6580DB7635E5",
  "MPParagraphStyle:5E96F392-D85B-4FC0-BE26-E3DB47947FD1",
  "MPTextStyling:E430B8EE-411D-4218-82E3-C30269D1803A",
  "MPParagraphStyle:74DD2FDA-6F5F-47C9-8862-23B34F4AC69B",
  "MPTextStyling:5507294C-AD8D-48EC-B86A-A4E5AE6C34D5",
  "MPParagraphStyle:776CD840-238C-40D2-A76D-90CC2F29DC78",
  "MPTextStyling:0A37ACFF-1D8A-4068-ABD0-0594C99C8EE0",
  "MPParagraphStyle:8EB89FAF-112B-4FBA-8E9A-CC393C1BDC1F",
  "MPTextStyling:055123F5-3704-49FE-B59B-D801598C67C4",
  "MPParagraphStyle:9FB73616-9404-4AB4-929E-4CCB2989E66A",
  "MPTextStyling:A3F0EF28-7A3F-4208-9205-5BD3FF8C1066",
  "MPParagraphStyle:C56C3DAC-C60D-47A6-A46C-33DD3D1D6A27",
  "MPTextStyling:2E79D76D-CAC1-492F-8709-DAD9FD369B8B",
  "MPParagraphStyle:D69F0A94-8608-4F62-9CBF-FF8C6A10526A",
  "MPTextStyling:6094A8B5-2751-489F-B4A5-699E54B57DC3",
  "MPParagraphStyle:D9CE7150-C0D6-4048-B3B1-4E0CF31607DB",
  "MPTextStyling:38ED2127-F6B6-4465-B739-09A3701A832D",
  "MPSection:6645426C-168C-4163-8853-E87E0E43AEF2",
  "MPSection:827D0408-E1E7-4764-B77E-4B498840FA2E",
  "MPSection:A049EEE4-0184-421A-8847-6F129FBD1A16",
  "MPSection:B354AF5D-0EA5-4BCC-881A-BA5BBE99A576",
  "MPSection:C07FEE6F-485C-44BF-B39B-B37D70847DCD",
  "MPSection:E9237AC2-9C70-4439-913B-31398AF0B681",
  "MPSection:EB4EF6B8-596E-4DD4-9BD3-142429EE5379",
  "MPStatusLabel:020444E9-396F-4E4E-8F89-92199F9B3888",
  "MPStatusLabel:18478729-CB92-41E4-ACDC-556FFCE3BE9F",
  "MPStatusLabel:4FD2C804-8396-4786-B442-B5676A7CEA8C",
  "MPTableStyle:AD84B382-664C-4960-A806-8ECC743B2284",
  "MPBorder:6BC92907-B7DF-4542-84CF-1456246EFEA9",
  "MPBorder:EB6A90B8-E464-4016-A6E7-90F704EC63C2",
  "MPBorder:EBB5D0B3-A9FE-49B3-ADAD-2CC5851D91F9",
  "MPBorder:2338AB83-D20B-4118-8C38-D0F330E85F23",
  "MPBorder:EC933CA4-ABBB-49D0-A2EF-4B0B2D1DA786",
  "MPKeywordsElement:DC752884-00A9-4A51-9764-FD2C6CEE1FEB",
  "MPManuscriptKeyword:1485E014-3119-4082-B51B-7370735D4EB9",
  "MPManuscriptKeyword:3C4E63A7-541A-4D5F-9332-4DC9EF3C9B2E",
  "MPParagraphElement:3EDC099C-3589-4B05-9D3A-22C3FE4BFB08",
  "MPListElement:83E77B5B-C99D-41CD-868E-C046B20B3E66",
  "MPListElement:0B92A2B7-F21D-47A4-B2D6-DB2650BC0744",
  "MPFigureElement:38658124-89B3-4D97-8FE2-DB0D5A579A16",
  "MPFigure:3DD0D874-52DF-4D0B-A39D-A12C19337438",
  "MPListing:2972554A-EDEF-4BDA-A2A2-7DC13138523D",
  "MPCitation:4A85057E-5E2F-4420-84A1-7C99D8D2D7BF",
  "MPCitationItem:49C67B5E-08A1-470E-AFA6-227B35C3A4C4",
  "MPCitationItem:4DADF651-183B-4CD7-9CE8-8B6A318BB830",
  "MPSection:03BAF068-5509-4FEF-A06E-FCCD74721842",
  "MPBibliographyElement:7AB68881-C996-4326-9A1E-FFF0D12B5E21",
  "MPBibliographyItem:03D92299-2B18-4265-B225-77EC2EC51036",
  "MPBibliographicName:2724979D-FC44-4973-83E0-F4BEFB4405A8",
  "MPBibliographicDate:6DE7F26D-AFB3-410C-828F-5C0B1A1FF3C4",
  "MPBibliographyItem:9168260C-0ADE-47F4-8AD0-1FB699B41CD0",
  "MPBibliographicName:3263B25F-E5DD-4DE8-A981-A9A9E3DED163",
  "MPBibliographicDate:434608C0-BB96-456E-9A2C-9A62BEB8F731",
  "MPTableElement:8C50999D-B1FF-4236-9E7B-8270BE1E1983",
  "MPTable:868B7CD4-3C60-44E9-9ADB-CF6B1E358D36",
  "MPListing:F414D820-B59A-4340-A862-1BD5CFC15361",
  "MPEquationElement:6C4AC91E-CE66-4393-9BC7-8AC3F5460647",
  "MPEquation:914B1CDA-93D1-4B06-BA7C-2B94FEF8C0A8",
  "MPListingElement:C0A14577-E556-4D26-B916-B14A38538C0B",
  "MPListing:26322D94-D5EA-47D7-87FE-4B3EFA4C0121",
]
`;
